### DISCUSSION TITLE
Notebooks: Remote Jupyter Server: Support custom auth schemas

### DISCUSSION THREAD

**Discussion by jschaeck** [microsoft/vscode-jupyter#10365] [New Ideas]
Hello, 

connecting VS Code notebooks to a remote server works really nice. However, right now , the expectation is that the authentication is exclusively done via ?token as a query param to the url.

Some cloud vendors hosting Jupyter Notebooks are using alternative auth schemas however and that prevents using VS Code NBs in that case. 

I wondered if support for additional auth schemas could be added. The jupyter community also did that in the Jupyter Server Gateway component: https://github.com/jupyter-server/jupyter_server/pull/529

As tokens might time out, it would be even better if it would be possible to register a callback function like getAuthHeaderForOutboundCall that users can implement on their own. (In there, the auth tokens could be recreated from an api key for example if they time out)

**Comment by rchiodo** [comment]
We do support user/password as well (common option when not using token for jupyter).

For other auth schemes there is an API that you could write an extension to support (and it supports timeout).

There's an example here of using it:
https://github.com/microsoft/vscode-jupyter/tree/main/src/test/datascience/extensionapi/exampleextension/ms-ai-tools-test

The Azure ML Extension in VS code uses it to provide connections to jupyter servers behind an Azure proxy.

**Comment by rchiodo** [comment]
The root of the API is the 'registerRemoteServerProvider' function. 
https://github.com/microsoft/vscode-jupyter/blob/11431255cd164c9488eb3e00b4255da02fa3f89a/src/platform/api.ts#L42

You call this from your extension.

You can see the implementation of the remote server provider in the example here:

https://github.com/microsoft/vscode-jupyter/blob/bd63750c91e6e23dacdfb99c24d88086e0d1acd4/src/test/datascience/extensionapi/exampleextension/ms-ai-tools-test/src/serverPicker.ts#L83

It is returning an object that has the authorization headers for the request and a timeout that causes the jupyter extension to reask later.
